Nonlin cal e2e test

hsergi opened the pull request:
## Describe your changes

I have followed the steps in [e2e testing](https://github.com/roman-corgi/corgidrp?tab=readme-ov-file#end-to-end-testing) to create an e2e test that starts with raw L1 data and derives a non-linearity calibration that is compared to the result derived with a similar tool and the same data during the TVAC tests.

## Type of change

Please delete options that are not relevant (and this line).

- New feature (non-breaking change which adds functionality): e2e test for a calibration type

## Reference any relevant issues (don't forget the #)

Tweaked from TVAC: changed OBSTYPE value from TVAC FITS files from ‘SCI’ to either ‘MNFRAME’ and ‘NONLIN’ as needed. ‘MNFRAME’ value was added to the wished Header Keywords in Confluence. 

## Checklist before requesting a review
- [x] I have linted my code (7.3/10, see bottom)
- [x] I have verified that all unit tests pass in a clean environment and added new unit tests, as needed
- [x] I have verified that all docstrings are properly formatted and added new documentation, as needed

Linting:
************* Module l1_to_l2a_e2e_nonlin
e2e_tests/l1_to_l2a_e2e_nonlin.py:65:21: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:95:11: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:101:0: R0914: Too many local variables (20/15) (too-many-locals)
e2e_tests/l1_to_l2a_e2e_nonlin.py:174:7: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:176:23: E1101: Instance of 'HDUList' has no 'data' member; maybe '_data'? (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:180:24: E1101: Instance of 'HDUList' has no 'data' member; maybe '_data'? (no-member)

------------------------------------------------------------------
Your code has been rated at 7.32/10

Commit:
Unnecessary import in test_kgain_cal and test_nonlin_cal

Commit:
Unnecessary import in test_kgain_cal and test_nonlin_cal

Commit:
Prepping calibrate_nonlin for corgidrp

# From Guillermo Gonzalez
# Nonlin 382: 51841-51870 (30), 51731-51840 (110), 51941-51984 (44),
# 51986-52051 (66), 55122-55187 (66), 55191-55256 (66)

# Additional information:
# L1 frame numbers for calibration frames collected during TV-20:
# Calibration product total num frames L1 range
# CIC map 80 50765-50796, 52125-52136, 56128-56133,
# 64348-64377
# Dark map 48 52077-52088, 56104-56109, 64228-64257

# Detailed information from Guillermo Gonzalez (ggonzo@tellus1.com)
# Frames 51841-51870 are used to generate the mean frame
# The EM gain values in the frame headers are 'commanded gain'. The values I
# listed in 'nonlin_table_240322.txt' are 'actual gain' values calculated
# from the EM gain calibration equation (Peter Williams can give you more details
# about that).
#
# norm_val=2020 DN
# NOTE: This and the following pixel values are from our Matlab script.
# To apply this to Python, subtract 1 from each.
# offset_colroi1=800 (-1)
# offset_colroi2=1000  (-1)
# offset_rowroi1=100  (-1)
# offset_rowroi2=1000  (-1)
# rowroi1=305  (-1)
# rowroi2=735  (-1)
# colroi1=1385  (-1)
# colroi2=1845  (-1)
# rowback11=20  (-1)
# rowback12=300  (-1)
# rowback21=740  (-1)
# rowback22=1000  (-1)
# colback11=1200  (-1)
# colback12=2000  (-1)
# colback21=1200  (-1)
# colback22= 2000  (-1)
# min_exp_time= 0,
# num_bins=didn't use this parameter in the original analysis (adopted default)
# min_bin= didn't use this parameter in the original analysis
# min_mask_factor= didn't use this parameter in the original analysis
# NOTE: I hard-wired 'min_mask' to be 1500 DN in the original analysis

Commit:
Prepping calibrate_nonlin for corgidrp

Commit:
Prepping calibrate_nonlin for corgidrp

Commit:
Prepping calibrate_nonlin for corgidrp

# From Guillermo Gonzalez
# Nonlin 382: 51841-51870 (30), 51731-51840 (110), 51941-51984 (44),
# 51986-52051 (66), 55122-55187 (66), 55191-55256 (66)

# Additional information:
# L1 frame numbers for calibration frames collected during TV-20:
# Calibration product total num frames L1 range
# CIC map 80 50765-50796, 52125-52136, 56128-56133,
# 64348-64377
# Dark map 48 52077-52088, 56104-56109, 64228-64257

# Detailed information from Guillermo Gonzalez (ggonzo@tellus1.com)
# Frames 51841-51870 are used to generate the mean frame
# The EM gain values in the frame headers are 'commanded gain'. The values I
# listed in 'nonlin_table_240322.txt' are 'actual gain' values calculated
# from the EM gain calibration equation (Peter Williams can give you more details
# about that).
#
# norm_val=2020 DN
# NOTE: This and the following pixel values are from our Matlab script.
# To apply this to Python, subtract 1 from each.
# offset_colroi1=800 (-1)
# offset_colroi2=1000  (-1)
# offset_rowroi1=100  (-1)
# offset_rowroi2=1000  (-1)
# rowroi1=305  (-1)
# rowroi2=735  (-1)
# colroi1=1385  (-1)
# colroi2=1845  (-1)
# rowback11=20  (-1)
# rowback12=300  (-1)
# rowback21=740  (-1)
# rowback22=1000  (-1)
# colback11=1200  (-1)
# colback12=2000  (-1)
# colback21=1200  (-1)
# colback22= 2000  (-1)
# min_exp_time= 0,
# num_bins=didn't use this parameter in the original analysis (adopted default)
# min_bin= didn't use this parameter in the original analysis
# min_mask_factor= didn't use this parameter in the original analysis
# NOTE: I hard-wired 'min_mask' to be 1500 DN in the original analysis

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

Commit:
Trying to decrease the differences with IIT

semaphoreP committed:
mocks should not access test_data dir

semaphoreP committed:
add filepath argument

Commit:
Adding NL recipe and e2e test

Commit:
Adding NL recipe and e2e test

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
working on walker

Commit:
working on walker

Commit:
working on walker

Commit:
working on walker

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

Commit:
OBSTYPE is in the primary header

semaphoreP committed:
Use ARRTYPE to determine detector areas

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Adding NL to walker

Commit:
Adding NL to walker

Commit:
Adding NL to walker

Commit:
Adding NL to walker

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
walker working

Now need to open results and compare with TVAC

Commit:
Comparison with TVAC

Commit:
Comparison with TVAC

Commit:
Comparison with TVAC

Commit:
Comparison with TVAC

Commit:
merging walker_with_dataset

Commit:
merging walker_with_dataset

Commit:
merging walker_with_dataset

Commit:
merging walker_with_dataset

Commit:
Add walker.py

Commit:
Add walker.py

Commit:
moved walker to corgidrp

Commit:
moved walker to corgidrp

Commit:
e2e test for NL calibration working (*)

I need to change some paths before submitting the PR

Commit:
e2e test for NL calibration working (*)

I need to change some paths before submitting the PR

Commit:
Merge branch 'nonlin_cal' of https://github.com/roman-corgi/corgidrp into nonlin_cal

Commit:
Adjusting default paths

Commit:
Do not track png figures

Commit:
run through pylint (7.3/10)

************* Module l1_to_l2a_e2e_nonlin
e2e_tests/l1_to_l2a_e2e_nonlin.py:65:21: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:95:11: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:101:0: R0914: Too many local variables (20/15) (too-many-locals)
e2e_tests/l1_to_l2a_e2e_nonlin.py:174:7: E1101: Instance of 'HDUList' has no 'header' member (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:176:23: E1101: Instance of 'HDUList' has no 'data' member; maybe '_data'? (no-member)
e2e_tests/l1_to_l2a_e2e_nonlin.py:180:24: E1101: Instance of 'HDUList' has no 'data' member; maybe '_data'? (no-member)

------------------------------------------------------------------
Your code has been rated at 7.32/10

Commit:
Fixing build errors

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

Commit:
Update l1_to_l2a_e2e_nonlin.py

semaphoreP commented on tests/e2e_tests/l1_to_l2a_e2e_nonlin.py:
rename this to something like `nonlin_e2e.py`. No need to include l1_to_l2a.

semaphoreP reviewed (changes requested):
Overall looks good. I just have some minor comments, and a few clarification questions. 

I also ran the code myself and the biggest memory use is actually in the prescan_biassub and detect_cosmic_ray steps. A separate issue (from this) to think about how to make them more memory efficient.

hsergi commented on corgidrp/calibrate_nonlin.py:
```diff
@@ -208,6 +208,9 @@ def calibrate_nonlin(dataset_nl,
       show_plot (bool): (Optional) display the plots. Default is False.
       verbose (bool): (Optional) display various diagnostic print messages.
         Default is False.
+      test_iit (bool) (Optional): If set to True, some structural changes are
```
Not using it: +/- 3% peak-to-peak differences in NL coefficients for a given DN and EM. Using it: +/-1 1% differences peak-to-peak.

hsergi commented on corgidrp/recipe_templates/l1_to_l2a_nonlin.json:
```diff
@@ -0,0 +1,34 @@
+        },
+        {
+            "name" : "calibrate_nonlin",
+            "keywords" : {
```
They are the ones used to derive the TVAC calibration table (with a different s/w: Matlab). Default ones are the ones in the delivery from the engineering version in June, though we do not have the NL table that results from it with TVAC data. It was used with sims. Using these values instead of the default values, brings the peak-to-peak differences in the NL table coefficients from +/-3% to +/-2%.

hsergi replied:
I have added the additional test_iit=true value to produce the smallest differences b/w the (Matlab) TVAC NL file and CORGI DRP's NL table (+/-1%). I'll use this for the assertion later on.
![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/890949b8-b03f-486e-88aa-48c9e5e7c514)

semaphoreP replied:
We should be writing these recipes with the idea that this is what we'll be running on the real science data. So, I would prefer the default values be the ones we plan to use for the real on-sky data. If you need to change them, we should change them in the test (E.g., by passing in a custom, modified template).

hsergi replied:
Agreed. That's what the template is doing. Scripts have default science values, not TVAC-comparison ones. I'll review it later today and confirm

hsergi replied:
These are values used for the comparison with the TVAC NL table only. I'll change the template to be the default one for science, and create a specific one for the e2e test.

semaphoreP replied:
Great. In that case, we probably want to have test that both recipes work.

hsergi replied:
Differences between/w the delivered code in June for NL (engineering version) and CORGI DRP are plotted below for different values of DN and EM

![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/39e23dc1-8780-491a-8e46-1f545214252c)

semaphoreP replied:
Differences of < 0.01% in nearly all cases seems pretty negligible to me. 

Do you have any thoughts to why there's still a bit of disagreement?

hsergi replied:
I'm trying to change the ROI in one step to match the one used by the engineering code. I may contact Kevin bc he thought that these changes wouldn't matter. I reported these kind of differences back then (end of July). Maybe there's something else? I compared Guillermo's NL table (run in Windows/Intel) with mine (MacOS/M2) and relative differences are <= 4e-15, so that's double precision (I spotted a different behavior in numpy sort() b/w both operating systems though that raised some errors and ~1% differences. That's understood and fixed in the above plot)

I'll keep you posted

hsergi replied:
I suggested some time ago to Marie that if there are (eventually) differences, say because bias subtraction is done at a different step in CORGI DRP (once and for all) compared to the engineering version, or else, we can always look at *the effect* of these differences in corrected data. if the differences when applying these tables are well below the error budget allowance for NL, then we'd be fine. Hopefully, we can reduce the differences too

hsergi replied:
Quick update: using the same ROI in CORGI DRP as in the ENG version (one column difference) reduces the differences a lot except for the case of unity gain. I also checked that the mean frame is exactly the same on both codes. I'll investigate more on the gain=1 data.

![CGI_EXCAM_L1_0000051731_NonLinearityCalibration](https://github.com/user-attachments/assets/45b1db9e-d17c-487a-b135-57e4c0aa64cb)

semaphoreP replied:
It's great to see that almost everything is agreeing. Kinda weird that EM=1 got worse... It seems like the first plot is actually better because everything is clearly in tolerance. 

You think the bias subtraction is done at a different step in corgidrp vs the engineering code? Can we move the prescan_biassub step in the recipe, or add a keyword toggle to prescan_biassub to match the engineering code?

What does ROI mean?

hsergi replied:
ROI means region of interest. I wrote a summary of the comparison in Confluence: https://collaboration.ipac.caltech.edu/display/romancoronagraph/Documenting+Calibration+Function+Porting

hsergi replied:
The decision made by DRP leads is to use the detector area from the engineering version of the code when deriving the NL correction table, instead of the one in detector_areas. This means to write the detector area in the NL recipe. I have committed the changes to comply with that, including cleaning up the e2e scripts that had other temporary options in consideration.

hsergi replied:
The PR should be good to be merged. P.S. I'll resolve the conflicts on 9/12. Today I have another urgent task to work on

hsergi commented on corgidrp/walker.py:
```diff
@@ -122,6 +124,14 @@ def guess_template(dataset):
+        elif image.pri_hdr['OBSTYPE'] == "MNFRAME":
+            # Disambiguate between NONLIN and KGAIN
+            for data in dataset:
+                if data.pri_hdr['OBSTYPE'] == "NONLIN":
```
That was my first intention. Sounds good for now. If Juergen is working on kgain, he'll add his version to these statements.

hsergi commented on corgidrp/mocks.py:
```diff
@@ -650,10 +650,14 @@ def make_fluxmap_image(
     random_array = np.random.normal(0, rn, (1200, 2200)) # e-
     # Generate random values from fluxmap from a Poisson distribution
     Poiss_noise_arr = emgain*np.random.poisson(time*f_map) # e-
+    ## Debugging
```
Unnecessary for the e2e test. Removed.

hsergi commented on tests/e2e_tests/l1_to_l2a_e2e_nonlin.py:
```diff
@@ -0,0 +1,210 @@
+    # We are going to make a new nonlinear calibration file using
+    # a combination of the II&T nonlinearty file and the mock headers from
+    # our unit test version of the NonLinearityCalibration
+    nonlin_dat = np.genfromtxt(tvac_caldir+'nonlin_table_240322.txt', delimiter=",")
```
It is used to create a FITS (CORGIDRP) file that is opened later on to have an analogous file/index structure when comparing the NL from the e2e and TVAC.

hsergi commented on tests/e2e_tests/l1_to_l2a_e2e_nonlin.py:
```diff
@@ -0,0 +1,210 @@
+            'format than the one from TVAC')   
+
+    rel_out_tvac_perc = 100*(nonlin_out_table[1:]/nonlin_tvac_table[1:]-1)
+    plt.figure(figsize=(10,6))
```
I have contacted Guillermo who has a better understanding of the impact of such differences. The max diff is 0.98% and the STD is 0.17%. If he agrees, I will use these values.

hsergi replied:
The 1% as a max difference b/w NL coefficients for pairs of DN, EM  is seen as reasonable by both Guillermo Gonzalez and Bijan Nemati.

Commit:
PR comments

Commit:
Assertion, eng recipe added

Commit:
Adding new NL table from TVAC

Commit:
row/col parameters matching eng version

There's a confusion in the initial exchange with Matlab about indices since Matlab starts with 1 instead of 0.

I have copied the values used by Guillermo Gonzalez in the eng version as shared on 9/9/24:

# offset column ROI range
offset_colroi1: 799
offset_colroi2: 1000

# ROI constants
rowroi1: 305
rowroi2: 736
colroi1: 1385
colroi2: 1846

# background ROIs
rowback11: 20
rowback12: 301
rowback21: 740
rowback22: 1001
colback11: 1200
colback12: 2001
colback21: 1200
colback22: 2001

Commit:
Update nonlin_e2e.py

Commit:
Update nonlin_e2e.py

Commit:
Comparison code bw Engineering and CORGI DRP

Differences are max: 0.04% and RMS: 0.003%

Commit:
Fine test b/w ENG and CORGI DRP

See the minor change in ROI

Commit:
test_iit not necessary any more

We are not comparing with the Matlab code used by Guillermo Gonzalez/tellus1 any more

Commit:
Update nonlin_e2e.py

Commit:
Implementing NL e2e decisions

Use same detector area as in the delivered version from "Calibrations" for the NL case

This is the one to be used:

        'prescan_reliable' : {
            'rows': 1200,
            'cols': 201,
            'r0c0': [0, 799]
            },

Commit:
--eng_version not needed any more

This option as meant to change some keyword values in the recipe at the time of running the e2e. The decision is to use the values in the default recipe, so no need to change them during the e2e tests

Commit:
Simplified recipe

Commit:
Current e2eversion

Commit:
pylint

Commit:
Final version of the NL calibration recipe for the PR#167

pylint thorws 7.3. Mostly due to some FITS HDU that may or may not have data

kjl0025 commented:
I looked over the Confluence page.  Sorry if I'm late to the game here.  I know there are a few differences in the ROI parameters, but they are fully customizable in the II&T code, right?  We could do a comparison between the DRP code and the II&T code with the exact same regions chosen (adapting the II&T parameters to match the defaults for the DRP), right?

semaphoreP commented:
They are, but the goal here is to have corgidrp reproduce the II&T code end-to-end (including pre-processing steps). So, we probably want to adjust corgidrp, rather than adjusting II&T, to show that corgidrp can produce the same results.

Commit:
Adding back CR detection and definite nonlin table

I removed it to speed up the e2e tests. Results are the same with or without CR detection with these TVAC data

Commit:
Merge branch 'main' into nonlin_cal

hsergi commented on tests/e2e_tests/l1_to_l2a_output/old/CGI_EXCAM_L1_0000051731_NonLinearityCalibration_macthedROI.png:
.gitignore had tests/e2e_tests/*/*.png not sure how it made it

Commit:
Delete CGI_EXCAM_L1_0000051731_NonLinearityCalibration_macthedROI.png

Commit:
Resolving one last conflict with detector.py

Commit:
Update l1_to_l2a_nonlin.json

semaphoreP committed:
change test keywords to match e2e testing keywords
